webroot/js/components/chat/chat-input.js

```diff
@@ -0,0 +1,281 @@
+    this.handlePaste = this.handlePaste.bind(this);
+
+    this.handleContentEditableChange = this.handleContentEditableChange.bind(this);
+  }
+
+  componentDidMount() {
+    this.getCustomEmojis();
+  }
+
+  getCustomEmojis() {
+    fetch(URL_CUSTOM_EMOJIS)
+      .then(response => {
+        if (!response.ok) {
+          throw new Error(`Network response was not ok ${response.ok}`);
+        }
+        return response.json();
+      })
+      .then(json => {
+        this.emojiPicker = new EmojiButton({
+          zIndex: 100,
+          theme: 'dark',
+          custom: json,
+          initialCategory: 'custom',
+          showPreview: false,
+          emojiSize: '30px',
+          // emojisPerRow: 6,
+          position: 'right-start',
+          strategy: 'absolute',
+        });
+        this.emojiPicker.on('emoji', emoji => {
+          this.handleEmojiSelected(emoji);
+        });
+      })
+      .catch(error => {
+        // this.handleNetworkingError(`Emoji Fetch: ${error}`);
+      });
+  }
+
+  handleEmojiButtonClick() {
+    if (this.emojiPicker) {
+      this.emojiPicker.togglePicker(this.emojiPickerButton.current);
+    }
+  }
+
+  handleEmojiSelected(emoji) {
+    const { inputHTML } = this.state;
+    let content = '';
+    if (emoji.url) {
+      const url = location.protocol + "//" + location.host + "/" + emoji.url;
+      const name = url.split('\\').pop().split('/').pop();
+      content = "<img class=\"emoji\" alt=\"" + name + "\" src=\"" + url + "\"/>";
+    } else {
+      content = emoji.emoji;
+    }
+
+    this.setState({
+      inputHTML: inputHTML + content,
+    });
+  }
+
+  // autocomplete user names
+	autoCompleteNames() {
+    const { chatUserNames } = this.props;
+		const { inputHTML } = this.state;
+		const position = getCaretPosition(this.formMessageInput.current);
+		const at = inputHTML.lastIndexOf('@', position - 1);
+		if (at === -1) {
+			return false;
+		}
+
+		let partial = inputHTML.substring(at + 1, position).trim();
+
+		if (partial === this.suggestion) {
+			partial = this.partial;
+		} else {
+			this.partial = partial;
+		}
+
+		const possibilities = chatUserNames.filter(function (username) {
+			return username.toLowerCase().startsWith(partial.toLowerCase());
+		});
+
+		if (this.completionIndex === undefined || ++this.completionIndex >= possibilities.length) {
+			this.completionIndex = 0;
+		}
+
+		if (possibilities.length > 0) {
+			this.suggestion = possibilities[this.completionIndex];
+
+      this.setState({
+        inputHTML: inputHTML.substring(0, at + 1) + this.suggestion + ' ' + inputHTML.substring(position),
+      })
+		}
+
+		return true;
+  }
+
+  handleMessageInputKeydown(event) {
+    const okCodes = [37,38,39,40,16,91,18,46,8];
+    const formField = this.formMessageInput.current;
+
+    let textValue = formField.innerText.trim(); // get this only to count chars
+
+    let numCharsLeft = this.maxMessageLength - textValue.length;
+		if (event.keyCode === 13) { // enter
```

Comment:
nit: since we're targeting modern browsers, you can avoid the cryptic `keyCode` numbers in favor of `event.code` if you want:

https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code

Reply:
Decided to use `event.key` instead. Didn't need the granularity of `event.code`.